### ISSUE TITLE
Overhaul completely the system of producing examples from songs during training to fix class imbalance and other problems

### ISSUE THREAD

**Issue by mayhemsloth** [mayhemsloth/Drum-Tabber#24]
Currently the training is done in the following way: all songs in the training set are loaded into a FullSet dataframe. To accomplish this, each song is processed through the Music Aligned Tab object, and then the resulting dataframe MAT is extracted and put into a dictionary, which is then used to create a multi-level dataframe whose top level is the song name. From this FullSet, a DataSet object is created, allowing the splitting of training/validation based on songs. During the actual training loop, each song is grabbed from the DataSet object, data augmentations are applied, and then input and target pairs are constructed and batched to give to the model for inference and backpropagation. 

There are some poor design choices here and I want to fix them. The main problem is related to **class imbalance**. Given any example in the FullSet, there's roughly a 10% chance that there's a drum onset event for any class. Some classes are more frequent (cymbals, bass, snare) and some classes are very infrequent (~1% for tom hits). I am currently handling this imbalance issue by using a weighted cross binary loss function, with weights calculated by the imbalance ratios of the FullSet, but that was always a lazy solution that I knew I would replace eventually. In a binary single class model, ideally you want to show your model as many positive examples as negative samples. In this way it isn't biased towards one outcome due to frequency alone. Drum-Tabber is a multi-label, multi-class classification problem, because any combination of drum event onsets can happen simultaneously (in practice only 4 can happen at once). The goal should not be to balance all potential *combinations* of classes, but mainly the relative presence of all individual classes with respect to each other *and* to fully negative examples. 

Another problem is that I don't randomize examples within a song. That is, the model is fed the *time ordered* examples. This should be easily fixed with any overhaul as long as I keep it in mind to design around it. 

Another problem is that the model experiences songs only one at a time, and not, for example, many song parts shuffled between different songs. and trained like that. For other reasons (like gathering statistics and loss based on song), I don't think I want to solve this problem. I also don't think it's that bad of a problem, especially if you end up doing 100+ epochs. Each song will tend to move the weights in some direction (in high-dimensional weight/loss space), and the fact that that movement is coordinated in time during training isn't that big of a deal. Ideally you would randomly choose between any example equally likely without replacement until all examples are exhausted. I'm okay with this not happening here.

Here's what I am thinking as a potential solution to this problem. 
1) During training, after DataSet yields a song, identify all the examples that have positive drum events labels (or time-keeping event labels as well). I will call this subset the "all positive" set. 
2) Statistics is done on the all positive set to find the relative presence of the classes. One class will be most frequent and its relevance ratio will be set to 1. Then more copies of examples of lower relevance ratio classes will be added to the all positive set until a certain threshold is hit (how to decide this threshold and which classes get what needs to be decided). This step is attempting to solve the problem with relative class imbalance via copying the minority classes. 
3) After the balanced all positive set is created, we still only have positive examples only and thus need to select negative examples. Assuming that the drum stem is accessible, there are multiple ways to approach this. For example, one way would be to take the balanced all positive set as the positive examples, and then create the negative examples by subtracting out the drum stem from the balanced all positive set. In this way, you would have exactly as many negative examples as positive for every class. Another example, again assuming that the drum stem is accessible, is to use the all balanced positive set as one set of positive examples, as well as the drum stem only version of that set, and then the drum-stem-subtracted set as negative examples. This would yield 2 positive examples for every negative one, which isn't exactly a good balance. To get another negative example, I would want to access the negative example that is "closest in time" (either backwards or forwards in time in the song) of each positive example. These negative samples would constitute the "most difficult" negative examples because they are right on the border of being a positive example (in time). I could randomly choose the backwards or forwards in time for each example, and then that would be 2 positive examples and 2 negative examples, creating a good balance. 
4) All statistics would be gathered and saved based only on the examples that are finally batched up, randomly shuffled and yielded to the model for inference and backpropagation. One advantage of this approach is that it should cut down on the *per song* training time because most of the song's fully negative examples (that is, those with no positive labels) are simply not batched and used in training. In any given song, around 75% of the song *is* fully negative examples. This may end up being a problem if the model doesn't have enough exposure to "easy" negative examples? But I doubt it. An easy fix for this would be to randomly choose some number of fully negative examples from the all negative set, which can be easily found as the complement of the all positive set. 

One potential problem of copying classes, or trying to ensure balance of positive vs. negative examples is that they interact with each other to a degree. For example, let's say you have 100 Snare drum events and 50 bass drum events and 25 cymbal events, and in this case all don't overlap at all. If you copy the bass drum events twice and the cymbal events 4 times, then there will be 100 Snare drum events and 200 *non* Snare drum events, and 100 bass drum events and 200 *non* bass drum events, and 100 cymbal events and 200 *non* cymbal events. We see that balancing the relative frequencies of the classes also, necessarily, introduces more negative class examples for every other class. This scales poorly with number of classes and with the total number of minority class examples. If we did this same exercise with 10 classes instead, we'd end up with the bass drum (and indeed every class) only being in 10% of all examples shown to the model, **even without the presence of fully negative examples**. I guess this is always a problem with multi-label classifications problems.

**This issue is resolved when the solution as described above is fully implemented.**
